fastrand.go

```diff
@@ -21,48 +20,65 @@ import (
 // is the concatenation of an initial seed and a 128-bit counter. Each time
 // the entropy is hashed, the counter is incremented.
 type randReader struct {
-	entropy []byte
-	h       hash.Hash
-	buf     []byte
-	mu      sync.Mutex
+	entropy chan []byte
 }
 
 // Read fills b with random data. It always returns len(b), nil.
 func (r *randReader) Read(b []byte) (int, error) {
-	r.mu.Lock()
 	n := 0
 	for n < len(b) {
-		// Increment counter.
-		*(*uint64)(unsafe.Pointer(&r.entropy[0]))++
-		if *(*uint64)(unsafe.Pointer(&r.entropy[0])) == 0 {
-			*(*uint64)(unsafe.Pointer(&r.entropy[8]))++
+		n += copy(b[n:], <-r.entropy)
+	}
+	return n, nil
+}
+
+// fillEntropy continuously fills r.entropy with new entropy.
+func (r *randReader) fillEntropy() {
+	// Create a hasher and fill it with 64 bytes of entropy. Technically only 16
+	// should be needed, but the underlying RNG may not be secure.
+	h := blake2b.New256()
+	_, err := io.CopyN(h, rand.Reader, 64)
+	if err != nil {
+		panic("fastrand: no entropy available")
+	}
+	seed := h.Sum(nil)
+
+	for {
+		for i := uint64(0); i < math.MaxUint64; i++ {
+			// Update the seed.
+			*(*uint64)(unsafe.Pointer(&seed[0])) = i
+
+			// Hash the seed.
+			h.Reset()
+			h.Write(seed)
+
+			// Send the entropy down the entropy channel.
+			r.entropy <- h.Sum(nil)
```

Comment:
This causes allocations and GC pressure because we continuously allocate new entropy, even though only a finite amount needs to exist at any given time (specifically, 32*1000). Profiling indicates that the impact isn't too severe though; around 1%.